src/semant/Semant.cpp

```diff
@@ -1605,13 +1605,49 @@ void Semant::transEvidence(absyn::Evidence* ne) {
 }
 
 void Semant::transQuery(absyn::Query* nq) {
+  // special check for for-loop in query statement
+  std::vector<std::shared_ptr<ir::VarDecl> > vds;
+  std::shared_ptr<ir::Expr> cond = nullptr;
+  if (nq->getVarDecls().size() > 0) { // has for loops
+    vds = transVarDecls(nq->getVarDecls(), true);
+    // add local variables
+    for (auto&v : vds) add_local_var(v);
```

Comment:
Let's check these local variables are all from types with fixed amount of instances.